pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/service/CookieUtils.scala

```diff
@@ -16,48 +25,39 @@ object CookieUtils {
       s"&expires=${authUser.expires}" +
       s"&multifactor=${authUser.multiFactor}"
 
-  private[service] def deserializeAuthenticatedUser(serializedForm: String): AuthenticatedUser = {
+  private[service] def deserializeAuthenticatedUser(serializedForm: String): Option[AuthenticatedUser] = {
     val data = serializedForm
       .split("&")
       .map(_.split("=", 2))
       .map{p => p(0) -> p(1)}
       .toMap
 
-    AuthenticatedUser(
-      user = User(data("firstName"), data("lastName"), data("email"), data.get("avatarUrl")),
-      authenticatingSystem = data("system"),
-      authenticatedIn = Set(data("authedIn").split(",").toSeq :_*),
-      expires = data("expires").toLong,
-      multiFactor = data("multifactor").toBoolean
+    for {
+      firstName <- data.get("firstName")
+      lastName <- data.get("lastName")
+      email <- data.get("email")
+      system <- data.get("system")
+      authedIn <- data.get("authedIn")
+      expires <- data.get("expires")
+      multifactor <- data.get("multifactor")
+    } yield AuthenticatedUser(
+      user = User(firstName, lastName, email, data.get("avatarUrl")),
+      authenticatingSystem = system,
+      authenticatedIn = Set(authedIn.split(",").toSeq :_*),
+      expires = expires.toLong,
+      multiFactor = multifactor.toBoolean
     )
   }
 
-  def generateCookieData(authUser: AuthenticatedUser, prvKey: PrivateKey): String = {
-    val data = serializeAuthenticatedUser(authUser)
-    val encodedData = new String(Base64.encodeBase64(data.getBytes("UTF-8")))
-    val signature = Crypto.signData(data.getBytes("UTF-8"), prvKey)
-    val encodedSignature = new String(Base64.encodeBase64(signature))
+  def generateCookieData(authUser: AuthenticatedUser, prvKey: PrivateKey): String =
+    CookiePayload.generateForPayloadText(serializeAuthenticatedUser(authUser), prvKey).asCookieText
 
-    s"$encodedData.$encodedSignature"
-  }
-
-  lazy val CookieRegEx = "^^([\\w\\W]*)\\.([\\w\\W]*)$".r
-
-  def parseCookieData(cookieString: String, pubKey: PublicKey): AuthenticatedUser = {
-
-    cookieString match {
-      case CookieRegEx(data, sig) =>
-        try {
-          if (Crypto.verifySignature(Base64.decodeBase64(data.getBytes("UTF-8")), Base64.decodeBase64(sig.getBytes("UTF-8")), pubKey)) {
-            deserializeAuthenticatedUser(new String(Base64.decodeBase64(data.getBytes("UTF-8"))))
-          } else {
-            throw new CookieSignatureInvalidException
-          }
-        } catch {
-          case e: SignatureException =>
-            throw new CookieSignatureInvalidException
-        }
-      case _ => throw new CookieParseException
-    }
-  }
+  // We would quite like to know, if a user is using an old (but accepted) key, *who* that user is- or to put it another
```

Comment:
I don't think I follow how this code does what the comment says..

From what I can see it'll either return an `AuthenticatedUser`, or a `SignatureNotValid` (or one of the other failure types), but none of these include the key that the user was using.. Am I missing something?

Reply:
Ah, but it [fails](https://github.com/guardian/pan-domain-authentication/actions/runs/10321653497/job/28575045968#step:6:21) because [we have to compile under Scala 2.12](https://github.com/guardian/pan-domain-authentication/issues/149):

```
[error] /home/runner/work/pan-domain-authentication/pan-domain-authentication/pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/service/CookieUtils.scala:41:48: value toLongOption is not a member of String
[error]       expires <- data.get("expires").flatMap(_.toLongOption)
[error]                                                ^
[error] /home/runner/work/pan-domain-authentication/pan-domain-authentication/pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/service/CookieUtils.scala:42:56: value toBooleanOption is not a member of String
[error]       multiFactor <- data.get("multifactor").flatMap(_.toBooleanOption)
[error]                                                        ^
[error] two errors found
```

...I'll make something Scala 2.12 compliant...